### ISSUE TITLE
Computed TypeMap cause panic in Terraform schema readMap when used with customdiff

### ISSUE THREAD

**Issue by ravitandonrt** [hashicorp/terraform#19978]
### Terraform Version

```
2019/01/11 10:53:31 [INFO] Terraform version: 0.11.7  41e50bd32a8825a84535e353c3674af8ce799161
2019/01/11 10:53:31 [INFO] Go runtime version: go1.10.1
```

### Terraform Configuration Files

```hcl
variable tenancy_ocid {}
variable user_ocid {}
variable fingerprint {}
variable private_key_path {}
variable compartment_ocid {}

provider "oci" {
  tenancy_ocid     = "${var.tenancy_ocid}"
  user_ocid        = "${var.user_ocid}"
  fingerprint      = "${var.fingerprint}"
  private_key_path = "${var.private_key_path}"
  region           = "us-ashburn-1"
}

locals {
  availability_domain = "HUYg:US-ASHBURN-AD-3"
  compartment_ocid    = "${var.compartment_ocid}"
  nat_offset          = "4"
}

# Network
resource "oci_core_vcn" "test_vcn" {
  compartment_id = "${local.compartment_ocid}"

  display_name = "test vcn"
  cidr_block   = "10.63.255.0/25"
}

resource "oci_core_internet_gateway" "ig" {
  compartment_id = "${local.compartment_ocid}"
  vcn_id         = "${oci_core_vcn.test_vcn.id}"
  display_name   = "Internet gateway"
}

resource "oci_core_subnet" "seed" {
  vcn_id              = "${oci_core_vcn.test_vcn.id}"
  compartment_id      = "${local.compartment_ocid}"
  availability_domain = "${local.availability_domain}"
  display_name        = "Seed"

  cidr_block = "${cidrsubnet(oci_core_vcn.test_vcn.cidr_block, 1, 0)}"

  prohibit_public_ip_on_vnic = "true"
}

# Compute
variable "services" {
  type = "list"
  default = [
    {
      name = "test1"
      outside_ip = "10.60.24.24"
    },
    {
      name = "test2"
      outside_ip = "10.60.24.48"
    }
  ]
}

data "null_data_source" "nat_entries" {
  count = "${length(var.services)}"

  inputs {
    name       = "${lookup(var.services[count.index], "name")}"
    inside_ip  = "${cidrhost(oci_core_subnet.seed.cidr_block, local.nat_offset + 1 + count.index)}"
    outside_ip = "${lookup(var.services[count.index], "outside_ip")}"
  }
}

resource "oci_core_instance" "nat" {
  compartment_id      = "${local.compartment_ocid}"
  availability_domain = "${local.availability_domain}"
  display_name        = "nat001"

  shape = "VM.Standard2.2"

  create_vnic_details {
    subnet_id              = "${oci_core_subnet.seed.id}"
    private_ip             = "${cidrhost(oci_core_subnet.seed.cidr_block, local.nat_offset)}"
    assign_public_ip       = "false"
    skip_source_dest_check = "true"
  }

  metadata {
    should_observe_dependency = "${jsonencode(data.null_data_source.nat_entries.*.outputs)}"
    this_should_also = "${oci_core_subnet.seed.time_created}"
    something = "something"
    another = "another"
  }
}
```

### Debug Output

```
2019/01/11 11:09:18 [DEBUG] Resource state not found for "oci_core_instance.nat": oci_core_instance.nat
2019/01/11 11:09:18 [DEBUG] ReferenceTransformer: "oci_core_instance.nat" references: []
2019/01/11 11:09:18 [TRACE] Graph after step *terraform.ReferenceTransformer:

oci_core_instance.nat - *terraform.NodePlannableResourceInstance
2019/01/11 11:09:18 [TRACE] Graph after step *terraform.RootTransformer:

oci_core_instance.nat - *terraform.NodePlannableResourceInstance
2019/01/11 11:09:18 [TRACE] dag/walk: added new vertex: "oci_core_instance.nat"
2019/01/11 11:09:18 [TRACE] dag/walk: walking "oci_core_instance.nat"
2019/01/11 11:09:18 [TRACE] vertex 'root.oci_core_instance.nat': walking
2019/01/11 11:09:18 [DEBUG] Interpolating computed map element attribute outputs (74D93920-ED26-11E3-AC10-0800200C9A66)
2019/01/11 11:09:18 [DEBUG] Interpolating computed map element attribute outputs (74D93920-ED26-11E3-AC10-0800200C9A66)
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci:
Error: Error running plan: 1 error(s) occurred:

* oci_core_instance.nat: 1 error(s) occurred:

* oci_core_instance.nat: unexpected EOF


panic: interface conversion: interface {} is nil, not map[string]interface {}
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci:
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: goroutine 44 [running]:
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*DiffFieldReader).readMap(0xc420380780, 0xc4207c3c10, 0x1, 0x1, 0xc4204caff0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/field_reader_diff.go:98 +0x717
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*DiffFieldReader).ReadField(0xc420380780, 0xc4207c3c10, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/field_reader_diff.go:71 +0x859
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*MultiLevelFieldReader).ReadFieldMerge(0xc4207c6f00, 0xc4207c3c10, 0x1, 0x1, 0x20fb2bc, 0x7, 0x0, 0x0, 0x0, 0x0, ...)
2019-01-11T11:09:18.421-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/field_reader_multi.go:45 +0x260
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceDiff).get(0xc42059f2c0, 0xc4207c3c10, 0x1, 0x1, 0x20fb2bc, 0x7, 0x0, 0x0, 0x0, 0x0, ...)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/resource_diff.go:485 +0xe3
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceDiff).getChange(0xc42059f2c0, 0x20fc49d, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/resource_diff.go:468 +0x340
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceDiff).GetChange(0xc42059f2c0, 0x20fc49d, 0x8, 0x0, 0xc420380840, 0x0, 0xc420380840)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/resource_diff.go:376 +0x9e
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/customdiff.ForceNewIfChange.func1(0xc42059f2c0, 0x1f096e0, 0xc420706360, 0xc4207c6ec0, 0x18050d9)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/customdiff/force_new.go:34 +0x54
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/customdiff.All.func1(0xc42059f2c0, 0x1f096e0, 0xc420706360, 0xc42037d820, 0xc42059f2c0)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/customdiff/compose.go:48 +0x9f
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.Diff(0xc42048d590, 0xc42060c0f0, 0xc4205494a0, 0xc4206e0500, 0x1f096e0, 0xc420706360, 0xc420377a40, 0x1d62906, 0x0)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:442 +0xb62
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Diff(0xc42041f0a0, 0xc42060c0f0, 0xc4205494a0, 0x1f096e0, 0xc420706360, 0x1, 0xc42063d0b0, 0xc42063d0b0)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:252 +0x187
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Diff(0xc4206173b0, 0xc42060c0a0, 0xc42060c0f0, 0xc4205494a0, 0x2fb8000, 0x0, 0x18)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:296 +0xa4
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Diff(0xc4206e0700, 0xc42037d5a0, 0xc42036e3b0, 0x0, 0x0)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:538 +0x57
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: reflect.Value.call(0xc4203509c0, 0xc4202e57e0, 0x13, 0x20f7b97, 0x4, 0xc420377f20, 0x3, 0x3, 0x0, 0x0, ...)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/usr/local/go/src/reflect/value.go:434 +0x905
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: reflect.Value.Call(0xc4203509c0, 0xc4202e57e0, 0x13, 0xc420498f20, 0x3, 0x3, 0x0, 0x0, 0x0)
2019-01-11T11:09:18.422-0800 [DEBUG] plugin.terraform-provider-oci: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2019-01-11T11:09:18.423-0800 [DEBUG] plugin.terraform-provider-oci: net/rpc.(*service).call(0xc420394e40, 0xc42008ae60, 0xc420271e30, 0xc42012ec00, 0xc4202e3b60, 0x1e1ffa0, 0xc42037d5a0, 0x16, 0x1e1ffe0, 0xc42036e3b0, ...)
2019-01-11T11:09:18.423-0800 [DEBUG] plugin.terraform-provider-oci: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2019-01-11T11:09:18.423-0800 [DEBUG] plugin.terraform-provider-oci: created by net/rpc.(*Server).ServeCodec
2019-01-11T11:09:18.423-0800 [DEBUG] plugin.terraform-provider-oci: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2019/01/11 11:09:18 [ERROR] root: eval: *terraform.EvalDiff, err: unexpected EOF
2019/01/11 11:09:18 [ERROR] root: eval: *terraform.EvalSequence, err: unexpected EOF
```

### Crash Output

https://gist.github.com/ravitandonrt/081de5e2c654fcf9b6d5f23b1c28e0f9

### Expected Behavior

`terraform plan` should succeed and show the map as computed

### Actual Behavior

`terraform plan` hits a panic during type conversion of interface{} nil to map[string]interface{}

### Steps to Reproduce

`terraform plan`

### Additional Context

**When does this issue occur:**
- If a field is configured with a customdiff that requires to call GetChange to execute forceNew - doesn't matter if it's ForceNewIf or ForceNewIfChange
- The field is a TypeMap with one or more nested values that are interpolated which make the entire map as computed

**NOTE:** This issue does not happen if there are no interpolations in TypeMap

**Where does this occur:**
- The panic occurs in file [`field_reader_diff.go` and function `readMap` at line 98](https://github.com/hashicorp/terraform/blob/master/helper/schema/field_reader_diff.go#L98), where the value of the field - if exists - is converted into a map
- The bug is likely in file [`helper/schema/field_reader_config.go` and function `readMap` at line 214](https://github.com/hashicorp/terraform/blob/master/helper/schema/field_reader_config.go#L214), where value is declared to be a interface{}, instead it should be declared as map[string]interface{} so that the calling function which expect a map can convert it safely. 
- In the same vein, we should also consider updating the same declaration at [`field_reader_diff.go` and function `readMap` at line 131](https://github.com/hashicorp/terraform/blob/master/helper/schema/field_reader_diff.go#L131)

### References

Original issue reported: https://github.com/terraform-providers/terraform-provider-oci/issues/685 - Right now, we reverted the customdiff in the said resource and are exploring the fix in vendor'ed Terraform schema as mentioned above.

We would appreciate the fix in both v11 and v12 of Terraform for TypeMap customdiff scenario.

The customdiff was introduced to address https://github.com/terraform-providers/terraform-provider-oci/issues/673

**Comment by alexng-canuck** [comment]
FYI @paultyng , here is the proposed fix for how we've addressed this:

[map.txt](https://github.com/hashicorp/terraform/files/3322139/map.txt)

**Comment by jbardin** [comment]
Hi @alexng-canuck,

I took the liberty of testing out the proposed changes, and the results were somewhat expected. Even seemingly minor changes to the sdk have unintended consequences. Purposely introducing a non-nil interface with a nil dynamic value is not something I would ever recommend in Go, and can definitely cause issues as it propagates through the code base, since it will subvert many `nil` checks. 

The fix I've proposed is more targeted to prevent the specific panics. Since the likelihood of these particular changes causing unintended behavior is quite low, we should be able to merge them even though the legacy sdk is essentially frozen at this point.
